collator/src/lib.rs

```diff
@@ -461,51 +512,52 @@ fn compute_targets(para_id: ParaId, session_keys: &[ValidatorId], roster: DutyRo
 /// build by the given `BuildParachainContext` and arguments to the underlying polkadot node.
 ///
 /// This function blocks until done.
-pub fn run_collator<P>(
-	build_parachain_context: P,
+pub fn run_collator<P, B>(
+	build_parachain: B,
 	para_id: ParaId,
 	key: Arc<CollatorPair>,
 	config: Configuration,
 ) -> polkadot_cli::error::Result<()> where
 	P: BuildParachainContext,
 	P::ParachainContext: Send + 'static,
 	<P::ParachainContext as ParachainContext>::ProduceCandidate: Send,
+	B: FnOnce() -> P,
```

Comment:
Don't you need this to be `Send`?

Reply:
It compiles in cumulus so I guess not. I think it helps that this is an FnOnce and not a Fn. For substrate it was a Fn and I believe it would have worked better with an FnOnce. (but in substrate it doesn't make sense to put an FnOnce there)